### DISCUSSION TITLE
pvlib development roadmap

### DISCUSSION THREAD

**Discussion by cwhanse** [pvlib/pvlib-python#1581] [Ideas]
@pvlib/pvlib-core and (at) everyone:

Several years back, @wholmgren and I proposed a [development roadmap](https://github.com/pvlib/pvlib-python/wiki/pvlib-python-Development-Roadmap) from v0.7. While perhaps too ambitious in scope, the roadmap helped focus efforts and many of those objectives have been achieved. Thanks to all who contributed with either code or comments - both are valued.

At the [2022 PVPMC workshop](https://pvpmc.sandia.gov/resources-and-events/events/) we heard from a group of users and contributors that it's time for a new roadmap. We intend the new roadmap to identify both short- and long-term development objectives. Short-term objectives are generally those for which no new information is needed; long-term objectives generally require a new approach or method to be identified or developed. We hope the roadmap will show where you can contribute in a meaningful way.

Our first task is to invite your comments on this draft roadmap. Please try to comment by 11/15/2022, and thanks for your input.

Short-term objectives:

1. Improve documentation pages. See discussions #329 #1546 #120 #60. This is a large task which could be broken into phases.
2. Address and close numerous issues related to IV curve calculations and functions: #411 #716 #516 #517 #418 #416 #1014 #959 #958 #1249 #1235 #1228 #1094 #1578 #1496 
3. Improve tutorials and documentation for the "I'm new to PV modeling and pvlib, how do I start?" audience.
4. Extend `pvlib.clearsky.detect_clearsky` to include the Bright-Sun algorithm #1061 and new work that addresses detecting clear-sky conditions in time-averaged and tilted-plane data ([preprint](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4249962)).
5. Add [Deline model](https://www.nrel.gov/docs/fy20osti/73541.pdf) for loss due to irradiance non-uniformity.
6. Add ohmic loss model. 
7. Add capability to read PAN and OND files.
8. Add capability to use SMARTS from pvlib.
9. Output ground irradiance from bifacial irradiance models.
10. Add LCOE calculations. Perhaps this is a long-term objective since nothing in pvlib currently has a financial unit.

Long-term objectives: 

1. Obtain pvlib.org
2. Add models for: DC-DC optimizer efficiency, inverter efficiency at non-unity power factor,
3. Migrate cell/module/string mismatch capability from [pvmismatch](https://github.com/SunPower/PVMismatch) into pvlib-python.
4. Add capability to adjust output for the time-resolution of the input data. For example, when using hourly irradiance it is common to compute solar position at the midpoint of each hour (or at the midpoint of the daylit portion of sunrise/sunset hours). There are also methods to adjust model output (hourly power/energy) to account for subhourly effects such as transitory inverter clipping.
5. Add capability to export model parameters and/or data in formats used by other modeling systems.
6. Extend detect-clearsky functions to label conditions with more resolution than clear/not clear; e.g., overcast, partially cloudy, etc.
11. Add snow loss models for bifacial and tracked systems.
12. Consider adding geographic and electrical coordinate systems. These coordinates would be needed if pvlib is to provide more detailed shading/bifacial/mismatch models.

**Comment by wholmgren** [comment]
1. Data model to support multiple inverters, integrating with ModelChain
2. https://github.com/pvlib/pvlib-python/issues/1269
3. https://github.com/pvlib/pvlib-python/issues/96, perhaps for now simply developing [openpvspec](https://github.com/BreakingBytes/openPVspec) and interfacing with pvlib

@cwhanse we have [pvsystem.dc_ohmic_losses](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.pvsystem.dc_ohmic_losses.html) so perhaps you mean to suggest https://github.com/pvlib/pvlib-python/issues/1069?

**Comment by adriesse** [comment]
It could be helpful to also try to articulate some higher level things like vision, mission, strategy and such.  The previous roadmap had little bits of this.

@cwhanse I seem to recall you were trying to get feedback specifically from industry users.  Did that provide any relevant insights?

**Comment by cwhanse** [comment]
@adriesse is that a vote to update the current roadmap wiki page? I'm OK to do that, although I don't know how much attention that page gets.

**Reply by adriesse** [reply]
No, just a suggestion to discuss these higher-level things.  They could influence the feature wish-list and prioritization.

**Comment by cwhanse** [comment]
I haven't gotten much specific feedback from industry users. Many I spoke with at RE+ don't use pvlib directly; they will use it to prototype or as a guide, then implement (often) in a different language.

**Reply by adriesse** [reply]
I was going to click on the sad emoiji, but that doesn't appear to be an option.  :(

**Reply by kurt-rhee** [reply]
As an industry user of pvlib my very selfish feedback on the pvlib roadmap posted above:

Short term objectives:
- No comments

Long term objectives:
- Migrating pvmismatch would be helpful for my team
- I have a set of functions that I have written which set the sun position depending on the context of what I am trying to calculate, for example I would like to midpoint of the timestamps when calculating what angle to use for true-tracking and I use the endpoint in order to determine what angles I need to use to avoid shading through the entire timestep.  Having the ability to calculate where in the timestep the calculation is occuring would make my code less verbose.
- I like SolarFarmer's approach to the JSON MNT file and would love for it to be implemented across the industry.

**Comment by kandersolar** [comment]
I have not expressed this publicly, but I would like to further develop the SAT capabilities of `pvlib.tracking`, including:
- Additional methods of determining the "ideal" rotation as alternatives to the astronomical method currently hardcoded in `singleaxis`.  One example is the optimization for diffuse conditions offered by several tracker manufacturers. 
- Additional methods of determining the backtracking position as alternatives to the simple `gcr` and `cross_axis_tilt` method currently implemented.  One example is the "split-boost" method for modules with half-cut cells as Aron Dobos discussed at the recent PVPMC meeting.  Another is terrain-based shade avoidance akin to NexTracker's TrueCapture which would require the new coordinate system and data model @cwhanse mentioned.
- Methods of modeling the real-world behavior of adjusting position in discrete steps along the lines of the discretized tracking gallery [example](https://pvlib-python.readthedocs.io/en/stable/gallery/solar-tracking/plot_discontinuous_tracking.html).  At least two discretization schemes are used by trackers today (constant time increment, constant angle increment).
- Miscellanea: systems with different "as-programmed" and "as-built" GCRs; wind stow logic; energy consumption of the tracker motors

Several of these items do not have any suitable reference in the literature that I'm aware of, so writing and publishing those would have to come first.  In addition to more clearly defining each step, some thought is needed about how best to combine them into a coherent modeling workflow.  A bonus objective is to develop this in coordination with the SAM team so that the two projects are modeling trackers consistently with each other.  @mjprilliman has expressed interest in this previously, although progress has stalled since then due to lack of time on my end.  Regardless this is on my personal aspirational roadmap for pvlib, so I bring it up here.

**Reply by kurt-rhee** [reply]
I would love to help with the terrain based shade avoidance section if helping does not conflict with the commercial interests of my employer.

**Comment by AdamRJensen** [comment]
An overview of the possible development for the iotools (including PR and issue references) can be found in this discussion: #1528

The main ideas are:
* Improve consistency (variable mapping and returning a tuple of data, meta)
* Add additional functions (e.g. for accessing ERA5, NASA POWER, horizon profile data)
* Somehow move to only having read/get functions (having the read function do the job of the parse function)
* A sub-page dedicated to introducing the iotools, providing an overview of what datasets and formats are available

**Comment by mikofski** [comment]
* #1501 arc glass
* #1178 off mpp inverter
* #1269 transformers

**Comment by adriesse** [comment]
Would it be useful to identify criteria for prioritization?  Or do feature ideas simply wait in the queue until a volunteer comes along?

**Reply by cwhanse** [reply]
> Would it be useful to identify criteria for prioritization? Or do feature ideas simply wait in the queue until a volunteer comes along?

Yes. My coarse setting: short-term means we have enough information and mostly need someone to work the issue; long-term means a desired feature that may require locating papers in literature or new work to be able to implement it.

As for prioritization - I haven't thought about sequencing these items or about sorting the short-term list by anticipated release version.

**Reply by adriesse** [reply]
How about expected benefit to different user groups as a criterion?  E.g. this capability is sorely needed by industry and no other software can do it yet, vs. commercial software already does this so pvlib should too, vs. this is a core capability for that area of research.

**Comment by mikofski** [comment]
seconding/emphasizing this issue as significantly important to me: Add models for: DC-DC optimizer efficiency, see #1588
